backend/app/services/pipeline_service.py

```diff
@@ -90,3 +94,59 @@ def delete_pipeline_by_id(self, pipeline_id: UUID) -> None:
                 raise ResourceInUseError(ResourceType.PIPELINE, str(pipeline_id), MSG_ERR_DELETE_RUNNING_PIPELINE)
             self._persistence.delete_by_id(pipeline_id, db)
             db.commit()
+
+    def get_pipeline_metrics(self, pipeline_id: UUID, duration_seconds: int = 60) -> PipelineMetrics:
+        """Calculate metrics for a pipeline over a specified time window."""
+        # First check if pipeline exists
+        pipeline = self.get_pipeline_by_id(pipeline_id)
+        if not pipeline.model_id:
+            raise ValueError("Cannot get metrics for a pipeline without an associated model.")
```

Comment:
I would change this check to `if pipeline.status != PipelineStatus.RUNNING` and raise a more specific exception. Then make sure to handle that exception in the endpoint (remap to 400 bad request).